Sofa/Component/Engine/Select/src/sofa/component/engine/select/SphereROI.inl

```diff
@@ -24,819 +24,218 @@
-                }
-            }
-        }
+        auto norm = sofa::helper::getWriteOnlyAccessor(d_normal);
+        norm.wref().normalize();
     }
 
-    //Tetrahedra
-    if (f_computeTetrahedra.getValue())
-    {
-        for(unsigned int i=0 ; i<tetrahedra.size() ; i++)
-        {
-            Tetra t = tetrahedra[i];
-            for (unsigned int j=0; j<cen.size(); ++j)
-            {
-                if (isTetrahedronInSphere(cen[j], rad[j], t))
-                {
-                    //tAngle > 0 ??
-                    tetrahedronIndices.push_back(i);
-                    tetrahedraInROI.push_back(t);
-                    break;
-                }
-            }
-        }
-    }
-
-    f_indices.endEdit();
-    f_edgeIndices.endEdit();
-    f_triangleIndices.endEdit();
-    f_quadIndices.endEdit();
-    f_tetrahedronIndices.endEdit();
+    return true;
 }
 
 template <class DataTypes>
-void SphereROI<DataTypes>::draw(const core::visual::VisualParams* vparams)
+void SphereROI<DataTypes>::roiDraw(const core::visual::VisualParams* vparams)
 {
-    if (!vparams->displayFlags().getShowBehaviorModels())
-        return;
-
-    const auto stateLifeCycle = vparams->drawTool()->makeStateLifeCycle();
-
-    const VecCoord* x0 = &f_X0.getValue();
     constexpr const sofa::type::RGBAColor& color = sofa::type::RGBAColor::cyan();
 
-    if(p_drawSphere.getValue()) // old classical drawing by points
-    {
-        ///draw the boxes
-        const type::vector<Vec3>& c=centers.getValue();
-        const type::vector<Real>& r=radii.getValue();        
-        std::vector<sofa::type::Vec3> drawcenters;
-        std::vector<float> drawradii;
-
-        for (unsigned int i=0; i<c.size() && i<r.size(); ++i)
-        {
-
-            drawcenters.push_back(c[i]);
-            drawradii.push_back(float(r[i]));
-            
-            if (edgeAngle.getValue() > 0)
-            {
-                vparams->drawTool()->drawCone(c[i], c[i] + direction.getValue()*(cos(edgeAngle.getValue()*M_PI / 180.0)*r[i]), 0, (float)sin(edgeAngle.getValue()*M_PI / 180.0)*((float)r[i]), color);
-            }
-
-            if (triAngle.getValue() > 0)
-            {
-                vparams->drawTool()->drawCone(c[i], c[i] + normal.getValue()*(cos(triAngle.getValue()*M_PI / 180.0)*r[i]), 0, (float)sin(triAngle.getValue()*M_PI / 180.0)*((float)r[i]), color);
-            }
-        }
-
-        vparams->drawTool()->setPolygonMode(0, true);
-        vparams->drawTool()->drawSpheres(drawcenters, drawradii, color);
-        vparams->drawTool()->setPolygonMode(0, false);
-    }
-
-    vparams->drawTool()->disableLighting();
-    std::vector<sofa::type::Vec3> vertices;
+    const auto& c = d_centers.getValue();
+    const auto& r = d_radii.getValue();
+    std::vector<sofa::type::Vec3> drawcenters;
+    std::vector<float> drawradii;
 
-    ///draw points in ROI
-    if( p_drawPoints.getValue())
+    for (unsigned int i=0; i<c.size() && i<r.size(); ++i)
```

Comment:
I have the impression there is a lot of xxx.getValue() in the loop. Should probably be factored out of the iterations.

Reply:
Right !